### ISSUE TITLE
Does Emscripten Supports ccache on Linux

### ISSUE THREAD

**Issue by Praveer1981** [emscripten-core/emscripten#11974]
Hi,
I want to know whether emscripten supports ccache on the Linux Platform ? If it supports then is there any specific flag which I need to set? I couldnt find any detail in the [Doc.](https://emscripten.org/)

**Comment by sbc100** [comment]
As far as I know ccache doesn't need any special support from the
compiler.   You just prepend the `ccache` command to the beginning of the
command line and you are done.

So

  $ emcc -c hello.c -o hello.o

becomes:

  $ ccache emcc -c hello.c -o hello.o

That seems to work for me anyway.

cheers,
sam

**Comment by Praveer1981** [comment]
ohh thanks @sbc100

**Comment by Praveer1981** [comment]
@sbc100 it  Does not work for me. I ma using 1.39.20 version of emscripten.However, 
If i run below command then it throws error 
 $ ccache emcc -c hello.c -o hello.o
Cleaned cache
ccache: error: missing equal sign in "hello.o"

However, I do not see any issue with the clang++
 $ ccache clang++ -c hello.c -o hello.o

**Comment by sbc100** [comment]
This works for me with ccache version 3.7.9.

```
$ ccache emcc -c ~/test/hello.c -o hello.o
$ file hello.o
hello.o: WebAssembly (wasm) binary module version 0x1 (MVP)
```

Does `emcc -c hello.c -o hello.o` work for you on its own?   

I have now idea what `missing equal sign in "hello.o"` means but it does't look a like an emscripten issue.

**Comment by sbc100** [comment]
Can you try with the latest emscripten (2.0.x)?  Or at least 1.40.1?

**Comment by Praveer1981** [comment]
@sbc100 yes, `emcc -c hello.c -o hello.o` works for me. based on my finding `missing equal sign in "hello.o"` error is due to wrong command.
$ ccache **-c** emcc main.cpp -o main.o

**Comment by Praveer1981** [comment]
@sbc100 I would also like to understand whether it works with CMake or not?

I can not see any way to use `ccache` in `Emscripten.cmake` file. to be more specific.
https://github.com/emscripten-core/emscripten/blob/master/cmake/Modules/Platform/Emscripten.cmake#L129

is there any document which I can refer to or is there any test which I can use as a reference for better understanding of using `ccache` with `emcc`

**Comment by sbc100** [comment]
As I say I don't think there is anything emscripten specific to do here.  If you just google for "using cache with cmake" you should be able to apply the advice you find.

For example : https://stackoverflow.com/questions/1815688/how-to-use-ccache-with-cmake

```
find_program(CCACHE_FOUND ccache)
if(CCACHE_FOUND)
    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache) # Less useful to do it for linking, see edit2
endif(CCACHE_FOUND)
```

But you will want to make sure `ccache emcc -c ~/test/hello.c -o hello.o` works first.. did I understand correctly that this is working for you now?

**Comment by Praveer1981** [comment]
@sbc100 yes you are right i do not see any error  when I use `ccache emcc -c ~/test/hello.c -o hello.o` .
As I mentioned above I can see that a `.ccache` directory has been created in the `home` directory. but I can not see any of the file which is related to `cache`. this(`ccache`) directory contains only one file `stats`. I have attached the snapshot of the stats `above`

However, if I use `ccache clang++ -c main.cpp -o main.o` I can see that it creats some cache related files in the `.ccache` directory. see the snapshot below:
![image](https://user-images.githubusercontent.com/24364597/92188661-c91bd580-ee8f-11ea-9109-9d8acbb5a0eb.png)

Since  I do not get this info using `emcc` So I afraid, if it really works with emscripten or not.

**Comment by Praveer1981** [comment]
When I use EMCC_DEBUG=2.
$  EMCC_DEBUG=2 ccache emcc -c main.cpp -o main.o
I do not see the `ccache` along with the `clang++`, that might be the reason why its not creating any of the cached file in the `.ccache` directory.

Below is the log info which was printed on the terminal
shared:DEBUG: successfully executed **/home/kumarpr/delete_it/Emscripten/emsdk/upstream/bin/clang++** -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=1 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -nostdsysteminc -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include/libcxx -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/lib/libcxxabi/include -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include/compat -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include/libc -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/lib/libc/musl/arch/emscripten -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/local/include -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include/SSE -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/lib/compiler-rt/include -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/lib/libunwind/include -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/cache/wasm/include -DEMSCRIPTEN -fignore-exceptions -c -o main.o main.cpp -Xclang -isystem/home/kumarpr/delete_it/Emscripten/emsdk/upstream/emscripten/system/include/SDL -c -o main.o -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr

**Comment by Praveer1981** [comment]
@sbc100 It looks like `ccache` does get into the emcc. so its not working as it was expected .

**Comment by sbc100** [comment]
If I run `ccache emcc hello.c -c` is certainly seems to work.

Each time I run it the percentage increases:

```
$ ccache -s
cache directory                     /home/sbc/.ccache
primary config                      /home/sbc/.ccache/ccache.conf
secondary config      (readonly)    /etc/ccache.conf
stats updated                       Mon Sep  7 04:33:52 2020
cache hit (direct)                     5
cache hit (preprocessed)               0
cache miss                             1
cache hit rate                     83.33 %
cleanups performed                     0
files in cache                         2
cache size                           8.2 kB
max cache size                       5.0 GB
```

We would not expect cache to show up in the internal running on clang++ from within emscripten.  

What does running your `ccache emcc -c main.cpp -o main.o` command do?  Does it work but just not change the hit rate?   Does it increase the cache miss count each time?

**Comment by Praveer1981** [comment]
If I run `ccache emcc -c main.cpp -o main.o`  
I do not see any error on the terminal. It creates a `.ccache` directory in my home directory.
Does it work but just not change the hit rate?  - yes
Does it increase the cache miss count each time? - No
Attach the snapshot 
![image](https://user-images.githubusercontent.com/24364597/92393118-ef709800-f151-11ea-83a7-a65df4aefef5.png)

**Comment by sbc100** [comment]
I wonder if the `compiler produced stdout` thing is the problem, does that number go up each time you run the `-c` command?  I don't see that in my output.

Does it produce anything on stdout when you run the `-c` command?

**Comment by Praveer1981** [comment]
`compiler produced stdout`  goes up each time I run the `-c` command.
I do not see any thing on my terminal when I run the `-c` command.

**Comment by sbc100** [comment]
Sounds like you might need to do a little debugging of ccache to see what its doing.   It certainly works for me.

**Comment by Praveer1981** [comment]
Sure, Let me try to debug ccache and try to investigate why its not working for me.

**Comment by jonassmedegaard** [comment]
I proposed to enable support for emscripten in the Debian ccache package, and the sceptical/cautious response from the ccache package maintainer might be helpful here: https://bugs.debian.org/972256

**Comment by sbc100** [comment]
Pretty much all of the emscripten-specific command line flags are link-time-only so are ignored in `-c` mode.    There are some `-s` option that effect codegen that will effect `-c`.   The `-s` options that effect compilation should me marks as such in the settings.js file.   We should probably highlight if there any non-s options that effect compile time (e.g. --default-obj-ext`)

Does cache actually require that symlink to work?  When I run `ccache emcc -c main.cpp -o main.o` it just works.    I guess that symlink tree is for folks that want to use ccache without actually adding it to the command line?

**Comment by jonassmedegaard** [comment]
Correct, the symlink is another way to use ccache.

Arguably it is not "necessary" to use the symlink, just as it is not "necessary" to use ccache at all.  It is a way to enable ccache at a higher abstraction level, e.g. in situations where it is unsuitable to micro-manage the path to the compiler.

**Comment by jrosdahl** [comment]
@Praveer1981 wrote:
> If i run below command then it throws error
> $ ccache emcc -c hello.c -o hello.o
> Cleaned cache
> ccache: error: missing equal sign in "hello.o"

You get the above output if you run `ccache -c emcc hello.c -o hello.o` instead of `ccache emcc -c hello.c -o hello.o` (note the different placements of `-c`). That's because `ccache -c` means "clean the cache" and `ccache -o x=y` means "set ccache configuration option x to y" (and ccache complains that you didn't provide an equal sign followed by a value).

> compiler produced stdout goes up each time I run the -c command.

Like indicated, that means that the compiler printed something to stdout, which is something ccache doesn't expect. Don't you get something printed if you run without ccache?

**Comment by Praveer1981** [comment]
thanks @jrosdahl and @sbc100.

**Comment by jonassmedegaard** [comment]
Seemingly the current state of affairs is that emsdk (not Emscripten generally) works with a ccache fork (not mainline ccache): https://github.com/emscripten-core/emscripten/pull/13681

**Comment by ericoporto** [comment]
Hi, is there any update on this? I would like to enable ccache support in our CI pipelines when using Emscripten, but am curious how is the support nowadays.
